charmhub_lp_tools/osci_sync.py

```diff
@@ -0,0 +1,247 @@
+                           project_vars: Dict[str, Any],
+                           auto_build_keys: List[Tuple[str, str, object]],
+                           ) -> Dict[str, str]:
+    """Generate a auto_build_channel compatible map.
+
+    Generate a auto_build_channel map based on the information available in
+    ``project_vars`` for the value referenced by ``osci_key``.
+
+    :param auto_build_channels: dictionary with the auto build channels to
+                                compare against.
+    :param project_vars: the project vars declared in osci.yaml
+    :param auto_build_keys: list of tuples where each tuple contains the key
+                            that should be accessed from the
+                            auto_build_channels map, the key that should be
+                            accessed from the project_vars map and finally a
+                            default value if the key is not found in
+                            project_vars.
+    :returns: A new dictionary with the updated values.
+
+    """
+    new_auto_build_channel = deepcopy(auto_build_channels)
+    for lp_key, osci_key, default in auto_build_keys:
+        new_value = project_vars.get(osci_key, default)
+        if auto_build_channels.get(lp_key) != new_value:
+            logger.debug('%s channel from %s to %s',
+                         lp_key,
+                         auto_build_channels.get(lp_key),
+                         new_value)
+            new_auto_build_channel[lp_key] = new_value
+
+    return new_auto_build_channel
+
+
+def setup_parser(subparser: argparse.ArgumentParser):
+    parser = subparser.add_parser(
+        'osci-sync',
+        help='Sync the config defined in osci.yaml to Launchpad.',
+    )
+    parser.add_argument(
+        '--i-really-mean-it',
+        dest='i_really_mean_it',
+        action='store_true',
+        default=False,
+        help=('This flag must be supplied to indicate that the sync/apply '
+              'command really should be used.'),
+    )
+    parser.add_argument(
+        '--repo-dir',
+        dest='repo_dir',
+        default=os.getcwd(),
+        metavar='DIR',
+        help=('Path to the git repository where the charm is located at, by '
+              'default it uses the current working directory.'),
+    )
+    parser.set_defaults(func=main)
+    return parser
+
+
+def main(args: argparse.Namespace,
+         gc: GroupConfig,
+         ) -> None:
+    logger.setLevel(getattr(logging, args.loglevel, 'ERROR'))
+    git_repo = git.Repo(args.repo_dir, search_parent_directories=True)
+    osci = load_osci_yaml(git_repo)
+    branch_name = get_branch_name(git_repo)
+    charm_name = get_charm_name(osci)
+    project_vars = get_project_vars(osci)
+    if not list(gc.projects(select=[charm_name])):
+        logger.error("No charm '%s' found; is the name correct?", charm_name)
+        sys.exit(1)
+
+    charm_project = list(gc.projects(select=[charm_name]))[0]
```

Comment:
It would be good to just evaluate `gc.projects(select=[charm_name]))` once.  It's a shame that Python doesn't have an `if let ...`.

I'd be inclined to use something like:

```python
try:
    charm_project = list(...)[0]
except IndexError:
    logger.error(...)
    sys.exit(1)
```

Reply:
this is a pattern already in use here - https://github.com/openstack-charmers/charmhub-lp-tools/blob/main/charmhub_lp_tools/main.py#L1210 .

Reply:
I will do the change, although I have no real preference on this ;-)